app_allocator/classes/dynamic_matrix_heuristic.py

```diff
@@ -0,0 +1,176 @@
+            judge_features = self.judge_features(judge)
+
+            needs_array = array([list(row.values()) for _, row
+                                 in self.application_needs.items()])
+            feature_weights_array = array([v for v in
+                                           self.feature_weights.values()])
+            needs_matrix = matrix(needs_array * feature_weights_array)
+            application_preferences = judge_features * needs_matrix.transpose()
+            apps = self.choose_n_applications(judge,
+                                              application_preferences,
+                                              available_batch_size)
+        for app in apps:
+            self.judge_assignments[judge].append(app)
+            self.judge_capacities[judge] -= 1
+            self._update_needs("assigned", judge, app)
+        return apps
+
+    def choose_n_applications(self, judge, application_preferences, n):
+        applications = (array(self.applications)[
+                (-application_preferences).argsort()]).flat
+        can_assign = self._can_assign_func(judge)
+        filtered = [app for app in applications if can_assign(app)]
+        return filtered[:n]
+
+    def _can_assign_func(self, judge):
+        '''returns a function which takes an application and returns True
+        if the application can be assigned to the judge.
+        Presumably this is where we'll put the z-score logic as well.
+        '''
+        def can_assign_to_judge(application):
+            return application not in self.judge_assignments[judge]
+
+        return can_assign_to_judge
+
+    def _feature_values(self, entity_sets):
+        return tuple(set([(feature, entity[feature.field])
+                          for entities in entity_sets
+                          for entity in entities
+                          for feature in self.features]))
+
+    def _feature_weights(self):
+        return {(feature.field, value):
+                feature.weight for feature, value in self.feature_values}
+
+    def initial_application_needs(self):
+        needs = OrderedDict()
+
+        for application in self.applications:
+            row = OrderedDict({(feature.field, value):
+                               feature.initial_need(application, value)
+                               for feature, value in self.feature_values})
+
+            needs[application] = row
+        return needs
+
+    def _update_needs(self, action, judge, application):
+        needs_dict = self.application_needs[application]
+        if action == "pass":
+            return
+        if action == "assigned":
+            assignments_list = self.judge_assignments[judge]
+            adjustment = ASSIGNED_VALUE
+        elif action == "finished":
+            assignments_list = self.completed_judge_assignments[judge]
+            adjustment = FINISHED_VALUE
+        assignments_list.append(application)
```

Comment:
I'd move the assignment of needs_dict down to here since it isn't used before this.